Gaget-tized the C++ header generator

JaredWright opened the pull request:
1) Added new gadgets to generate accessor functions, namespaces, and
dependencies

2) Re-wrote the C++ header generator to use gadgets

Signed-off-by: JaredWright <jared.wright12@gmail.com>

JaredWright commented:
So, [this](https://github.com/Bareflank/shoulder/blob/20fda1405d1e88d004628b1796951292e33210b6/shoulder/generator/cxx_header_generator.py) is what a generator ends up looking like using the decorator-based gadget idea (Issue #18) to its full potential.

What do you think @JWZepf ? Did this end up better or worse than before?

JWZepf commented on shoulder/generator/cxx_header_generator.py:
```diff
@@ -48,350 +46,403 @@ def generate(self, objects, outpath):
+        """
+        Generate accessors for a single register
+        """
+        self._generate_register_comment(outfile, reg)
```
I like moving these long print formatting blocks into separate functions. A lot easier to read and a lot easier to edit

JWZepf commented on shoulder/generator/cxx_header_generator.py:
```diff
@@ -48,350 +46,403 @@ def generate(self, objects, outpath):
             for idx, fieldset in enumerate(fieldsets):
-                outfile.write("\n")
-
-                if(fieldset.condition is not None):
```
We lost this fieldset condition comment block and I think that is a useful feature to have in the header. Maybe put this block back in or create a gadget that can append fieldsets with their condition description.

JaredWright replied:
Good catch, this got removed by accident. Will add back

JWZepf commented on shoulder/generator/cxx_header_generator.py:
```diff
@@ -48,350 +46,403 @@ def generate(self, objects, outpath):
-    def _generate_field_accessors(self, reg, field, outfile):
+            _in_field_namespace(self, outfile)
+
+    def _generate_field_constants(self, outfile, reg, field):
```
I'm a big fan of using these constant definitions instead of hardcoding it into the function. It makes it very easy to go to that field and know the critical information for it quickly.

JaredWright commented on include/regs.h:
```diff
@@ -1,27 +1,5 @@
-//
```
I removed it to make copy/pasting into a generated files easier, but I'll add it back in favor of installing this file along side generated files.

JaredWright commented on shoulder/gadget/accessor_macros.py:
```diff
@@ -0,0 +1,51 @@
+    used as the implementation for C/C++ register access functions
+
+    Args:
+        path (str): Path to a regstier access file
```
Good catch

JaredWright commented on shoulder/gadget/accessor_macros.py:
```diff
@@ -0,0 +1,51 @@
+#
```
Agreed, will remove this file.

JaredWright commented:
Ok @JWZepf, I believe that I addressed all of your comments. Take a look through my follow up commit, and then I can squash it if you're happy.

As we discussed on Friday, I renamed "regs.h" to "aarch64_gcc_accessor_macros.h". This file now gets installed alongside all generated files and ```#include```'ed in, instead of copy/pasting it into generated files.

JWZepf commented:
Ok, this all looks great to me! Merging 👍
